scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+        The psrfits main header object
+    subint_hdu : HDU oject
+        The psrfits data HDU.
+
+    Note
+    ----
+    Right now we are assuming the data rows are continuous in time and the
+    frequency are the same.
+    """
+
+    _properties = ('start_time', 'sample_rate', 'shape', 'samples_per_frame',
+                   'polarization', 'frequency')
+    _header_defaults = subint_header
+    _req_columns = subint_columns
+
+    def __init__(self, header_hdu, subint_hdu=None):
+        self.header_hdu = header_hdu
+        if subint_hdu is None:
+            super(SubintHDU, self).__init__(name='SUBINT')
+            self._init_empty()
+        else:
+            super(SubintHDU, self).__init__(header=subint_hdu.header,
+                                            data=subint_hdu.data)
+        self.verify()
+        self.offset = 0
+
+    def _init_empty(self):
+        for k in self._header_defaults:
+            self.header.set(k[0], k[1], k[2])
+
+    def make_columns(self, name):
+        cols = []
+        for cl in self._columns_defaults:
+            col = fits.Column()
+
+    def verify(self):
+        if self.header['EXTNAME'].replace(' ', '') != "SUBINT":
+            raise ValueError("Input HDU is not a SUBINT type.")
+
+    @property
+    def mode(self):
+        return self.header_hdu.obs_mode
+
+    @property
+    def sample_rate(self):
+        # We assume the sample rate are the same for all the rows.
+        sample_time = u.Quantity(self.data[0]['TSUBINT'] /
+                                 self.samples_per_frame, u.s)
+        return 1.0 / sample_time
+
+    @sample_rate.setter
+    def sample_rate(self, val):
+        if hasattr(val, 'unit'):
+            val = val.to(u.Hz).value
+        self.header['TBIN'] = 1.0 / val
+
+    @property
+    def start_time(self):
+        # NOTE should we get the start time for each raw, in case the time gaps
+        # in between the rows
+        file_start = self.header_hdu.start_time
+        if "OFFS_SUB" in self.columns.names:
+            subint_times = u.Quantity(self.data['OFFS_SUB'], u.s, copy=False)
+            sample_time = 1.0 / self.sample_rate
+            start_time = (file_start + subint_times[0] -
+                          self.samples_per_frame / 2 * sample_time)
+        else:
+            start_time = file_start
+        return start_time
+
+    @start_time.setter
+    def start_time(self, time):
+        # NOTE this sets the start time of the HDU, not the file start time.
+        dt = (time - self.header_hdu.start_time).to(u.s)
+        center_off = dt + 1.0 / self.sample_rate * self.samples_per_frame / 2
+        if "OFFS_SUB" in self.columns.names:
+            first_off = self.data['OFFS_SUB'][0]
+            self.data['OFFS_SUB'] = (self.data['OFFS_SUB'] - first_off +
+                                     center_off)
+        else:
+            self._req_columns['OFFS_SUB']['value'] = center_off
+
+    @property
+    def nrow(self):
```

Comment:
`nrow`, `npol`, and `nbin` seem implementation details; do we really need them?

Reply:
I was trying to avoid typing the same thing. The other properties (raw_shape, data_shape, and sample_per_frame) use them a lot.  I like the way those numbers are represented by a variable other than an index in the shape property.

Reply:
I perhaps am misunderstanding the class structure - I think it would be logical for the HDU wrapper to have this, but not the reader. Maybe that is in fact the case?